ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -332,6 +346,48 @@ initBlockFetchConsensusInterface cfg chainDB getCandidates blockFetchSize btime
     readFetchedMaxSlotNo :: STM m MaxSlotNo
     readFetchedMaxSlotNo = ChainDB.getMaxSlotNo chainDB
 
+    filterCandidates ::
+         AnchoredFragment (Header blk)
+      -> [CandidateFragment (Header blk)]
+      -> [Either CandidateDecline (ChainSuffix (Header blk))]
+    filterCandidates ours = map $ \cand ->
+        let CandidateFragment {
+                candidateChain
+              , candidateIsLowDensity
+              } = cand
+        in
+        if not $
+                plausibleCandidateChain ours candidateChain
+             || candidateIsLowDensity
+        then Left DeclineNotPlausible
+        else case mkChainSuffix ours candidateChain of
+                Right suffix                   ->
+                  Right suffix
+                Left ChainSuffixNoIntersection ->
+                  Left DeclineStale
+                Left ChainSuffixEmpty          ->
+                  -- If the suffix is empty, it means the candidate chain was
+                  -- equal to the current chain and didn't fork off. Such a
+                  -- candidate chain is not a plausible candidate, so it must
+                  -- have been filtered out. /Unless/ it had been marked as
+                  -- plausible; this can happen eg due to the
```

Comment:
"In the meantime, the blocks that ChainSync marked as necessary (ie low-density) are now on our current chain."